src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1320 @@
+   /// operations (printed to stderr). You can also set a constant
+   /// BSON_DSL_DEBUG within the scope of a DSL command to selectively debug
+   /// only the commands within that scope.
+   BSON_DSL_DEBUG = 0
+};
+
+#ifdef _WIN32
+#define _bson_thread_local_comdat __declspec(thread) __declspec(selectany)
+#else
+#define _bson_thread_local_comdat __attribute__ ((weak)) __thread
+#endif
+
+#ifdef __GNUC__
+// GCC has a bug handling pragma statements that disable warnings within complex
+// nested macro expansions. If we're GCC, just disable -Wshadow outright:
+BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic ignored \"-Wshadow\""))
+#endif
+
+#define _bsonDSL_disableWarnings()                     \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic push");) \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic ignored \"-Wshadow\"");)
+
+#define _bsonDSL_restoreWarnings() \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic pop");)
+
+/**
+ * @brief Parse the given BSON document.
+ *
+ * @param doc A bson_t object to walk. (Not a pointer)
+ */
+#define bsonParse(Document, ...)                              \
+   _bsonDSL_begin ("bsonParse(%s)", _bsonDSL_str (Document)); \
+   _bsonDSL_disableWarnings ();                               \
+   bool _bvHalt = false;                                      \
+   bsonParseError = NULL;                                     \
+   const bool _bvContinue = false;                            \
+   const bool _bvBreak = false;                               \
+   (void) _bvContinue;                                        \
+   (void) _bvBreak;                                           \
+   (void) _bvHalt;                                            \
+   _bsonDSL_eval (_bsonParse ((Document), __VA_ARGS__));      \
+   _bsonDSL_restoreWarnings ();                               \
+   _bsonDSL_end
+
+/**
+ * @brief Visit each element of a BSON document
+ */
+#define bsonVisitEach(Document, ...)                              \
+   _bsonDSL_begin ("bsonVisitEach(%s)", _bsonDSL_str (Document)); \
+   _bsonDSL_disableWarnings ();                                   \
+   bool _bvHalt = false;                                          \
+   (void) _bvHalt;                                                \
+   _bsonDSL_eval (_bsonVisitEach ((Document), __VA_ARGS__));      \
+   _bsonDSL_restoreWarnings ();                                   \
+   _bsonDSL_end
+
+#ifndef __INTELLISENSE__
+#define bsonBuildContext (*_bsonBuildContextThreadLocalPtr)
+#define bsonVisitContext (*_bsonVisitContextThreadLocalPtr)
+#define bsonVisitIter (bsonVisitContext.iter)
+#define bsonParseIter (bsonVisitContext.iter)
+#else
+/// The context of the current build operation
+extern struct _bsonBuildContext_t bsonBuildContext;
+/// The context of the current parse/visit operation
+extern struct _bsonVisitContext_t bsonVisitContext;
+/// An iterator pointing to the current BSON value being visited/parsed.
+extern bson_iter_t bsonVisitIter, bsonParseIter;
+#endif
```

Comment:
What is the purpose of this block (why `__INTELLISENSE__`)?

Reply:
It's to "lie" to the IDE, only to hide the macro expansion-on-hover, but that's getting really pedantic. Probably not worth the added complexity.